libs/Helpers/src/Helpers/QObjectVectorModel.h

```diff
@@ -0,0 +1,86 @@
+#pragma once
+#include <QAbstractListModel>
+#include <QQmlEngine>
+
+namespace Status::Helpers {
+
+/// Generic typed QObject provider model
+///
+/// Supports: source model update
+/// \todo rename it to SharedQObjectVectorModel
+template<typename T>
+class QObjectVectorModel final : public QAbstractListModel
+{
+    static_assert(std::is_base_of<QObject, T>::value, "Template parameter (T) not a QObject");
+
+public:
+
+    using ObjectContainer = std::vector<std::shared_ptr<T>>;
+
+    explicit QObjectVectorModel(ObjectContainer initialObjects, const char* objectRoleName, QObject* parent = nullptr)
+        : QAbstractListModel(parent)
+        , m_objects(std::move(initialObjects))
+        , m_roleName(objectRoleName)
+    {
+    }
+    explicit QObjectVectorModel(const char* objectRoleName, QObject* parent = nullptr)
+        : QObjectVectorModel(ObjectContainer{}, objectRoleName, parent)
+    {}
+    ~QObjectVectorModel() {};
+
+    QHash<int, QByteArray> roleNames() const override {
+        return {{ObjectRole, m_roleName}};
+    };
+
+    virtual int rowCount(const QModelIndex& parent = QModelIndex()) const override {
+        Q_UNUSED(parent)
+        return m_objects.size();
+    }
+
+    virtual QVariant data(const QModelIndex& index, int role) const override {
+        if(!QAbstractItemModel::checkIndex(index) || role != ObjectRole)
+            return QVariant();
+
+        return QVariant::fromValue<QObject*>(m_objects[index.row()].get());
+    }
+
+    const T* at(size_t pos) const { return m_objects.at(pos).get(); };
+    std::shared_ptr<T> get(size_t pos) { return m_objects.at(pos); };
+    size_t size() const { return m_objects.size(); };
+    void clear() { m_objects.clear(); };
+
+    void push_back(std::shared_ptr<T> newValue) {
+        beginInsertRows(QModelIndex(), m_objects.size(), m_objects.size());
+        m_objects.push_back(newValue);
+        endInsertRows();
+    };
+
+    void resize(size_t count) {
+        if(count > m_objects.size()) {
+            beginInsertRows(QModelIndex(), m_objects.size(), count - 1);
+            m_objects.resize(count);
+            endInsertRows();
+        }
+        else if(count < m_objects.size()) {
+            beginRemoveRows(QModelIndex(), count, m_objects.size() - 1);
+            m_objects.resize(count);
+            endRemoveRows();
+        }
+    };
+
+    void set(size_t pos, std::shared_ptr<T> newVal) {
+        m_objects.at(pos) = newVal;
+        emit dataChanged(index(pos, 0), index(pos, 0), {});
```

Comment:
Just `index(row)` the column is kinda not needed in a QAbstractListModel (https://doc.qt.io/qt-5/qabstractlistmodel.html#index)